src/unix/udp.c

```diff
@@ -576,6 +587,108 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
 }
 
 
+static int uv__udp_set_source_membership4(uv_udp_t* handle,
+                                          const struct sockaddr_in* multicast_addr,
+                                          const char* interface_addr,
+                                          const struct sockaddr_in* source_addr,
+                                          uv_membership membership) {
+  struct ip_mreq_source mreq;
+  int optname;
+  int err;
+
+  err = uv__udp_maybe_deferred_bind(handle, AF_INET, UV_UDP_REUSEADDR);
+  if (err)
+    return err;
+
+  memset(&mreq, 0, sizeof(mreq));
+
+  if (interface_addr) {
+    err = uv_inet_pton(AF_INET, interface_addr, &mreq.imr_interface.s_addr);
+    if (err)
+      return err;
+  } else {
+    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+  }
+
+  mreq.imr_multiaddr.s_addr = multicast_addr->sin_addr.s_addr;
+  mreq.imr_sourceaddr.s_addr = source_addr->sin_addr.s_addr;
+
+  switch (membership) {
```

Comment:
I would just write this as an `if...else if...else`. It would be several lines shorter, and there are only a couple of options. I'd also move it to the top of the function so that `-EINVAL` can be returned right away if necessary.

Reply:
I agree with you, but in this case functions `uv_udp_set_membership` and `uv_udp_set_source_membership` would be written in different manner. If this OK, I would replace `switch .. case` with `if .. else` construction.

Reply:
Yea, I'd go with `if...else`

Reply:
done